File: src/NovelRT/Ecs/UnsafeComponentView.cpp
// Copyright © Matt Jones and Contributors. Licensed under the MIT Licence (MIT). See LICENCE.md in the repository root
// for more information.

#include <NovelRT/Ecs/UnsafeComponentView.h>
#include <utility>

namespace NovelRT::Ecs
{
    UnsafeComponentView::UnsafeComponentView(size_t poolId,
                                             std::shared_ptr<ComponentBufferMemoryContainer> memoryContainer) noexcept
        : _poolId(poolId), _memoryContainer(std::move(memoryContainer))
    {
    }

    void UnsafeComponentView::PushComponentUpdateInstruction(EntityId entity, void* instructionState)
    {
        _memoryContainer->PushComponentUpdateInstruction(_poolId, entity, instructionState);
    }

    void UnsafeComponentView::RemoveComponent(EntityId entity)
    {
        _memoryContainer->PushComponentUpdateInstruction(_poolId, entity, _memoryContainer->GetDeleteInstructionState().GetDataHandle());
    }

    ComponentBufferMemoryContainer::ImmutableDataView UnsafeComponentView::GetComponent(EntityId entity) const
    {
        return _memoryContainer->GetComponent(entity);
    }

    ComponentBufferMemoryContainer::ImmutableDataView UnsafeComponentView::GetComponentUnsafe(
        EntityId entity) const noexcept
    {
        return _memoryContainer->GetComponentUnsafe(entity);
    }

    size_t UnsafeComponentView::GetImmutableDataLength() const noexcept
    {
        return _memoryContainer->GetImmutableDataLength();
    }

    // clang-format off

    SparseSetMemoryContainer::ConstIterator UnsafeComponentView::begin() const noexcept
    {
        return _memoryContainer->begin();
    }

    SparseSetMemoryContainer::ConstIterator UnsafeComponentView::end() const noexcept
    {
        return _memoryContainer->end();
    }

    // clang-format on
} // namespace NovelRT::Ecs

File: src/NovelRT.DotNet/Interop/Ecs/UnsafeComponentView.cs
// Copyright © Matt Jones and Contributors. Licensed under the MIT License (MIT). See LICENCE.md in the repository root for more information.

using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace NovelRT.Interop.Ecs
{
    public class UnsafeComponentView : ShimObject
    {
        public UnsafeComponentView(nuint poolId, ComponentBufferMemoryContainer container)
        {
            internalHandle = NovelRT.Nrt_UnsafeComponentView_Create(poolId, container.internalHandle);
        }
        public UnsafeComponentView(nuint poolId){ throw new NotImplementedException(); }
        public UnsafeComponentView(ComponentBufferMemoryContainer container) { throw new NotImplementedException();  }
        public UnsafeComponentView(IntPtr viewId){ throw new NotImplementedException(); }
    }
}
